internal/cp/proxy_factory.go

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cp
+
+import (
+	"context"
+	"github.com/hazelcast/hazelcast-go-client/internal/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/cp/types"
+	"github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
+	"github.com/hazelcast/hazelcast-go-client/internal/invocation"
+	"github.com/hazelcast/hazelcast-go-client/internal/logger"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto/codec"
+	iserialization "github.com/hazelcast/hazelcast-go-client/internal/serialization"
+	"strings"
+	"time"
+)
+
+const (
+	atomicLongService      = "hz:raft:atomicLongService"
+	atomicReferenceService = "hz:raft:atomicRefService"
+	countDownLatchService  = "hz:raft:countDownLatchService"
+	lockService            = "hz:raft:lockService"
+	semaphoreService       = "hz:raft:semaphoreService"
+
+	defaultGroupName    = "default"
+	metadataCpGroupName = "metadata"
+)
+
+type bundle struct {
+	invocationService    *invocation.Service
+	serializationService *iserialization.Service
+	invocationFactory    *cluster.ConnectionInvocationFactory
+	logger               *logger.LogAdaptor
+}
+
+func (b bundle) check() {
+	if b.invocationService == nil {
+		panic("invocationFactory is nil")
+	}
+	if b.invocationFactory == nil {
+		panic("invocationService is nil")
+	}
+	if b.serializationService == nil {
+		panic("serializationService is nil")
+	}
+	if b.logger == nil {
+		panic("logger is nil")
+	}
+}
+
+type proxyFactory struct {
+	bundle *bundle
+}
+
+func newProxyFactory(ss *iserialization.Service, cif *cluster.ConnectionInvocationFactory, is *invocation.Service, l *logger.LogAdaptor) *proxyFactory {
+	b := &bundle{
+		invocationService:    is,
+		invocationFactory:    cif,
+		serializationService: ss,
+		logger:               l,
+	}
+	b.check()
+	return &proxyFactory{
+		bundle: b,
+	}
+}
+
+func (m *proxyFactory) getOrCreateProxy(ctx context.Context, sn string, n string) (interface{}, error) {
+	var (
+		err error
+		pxy string
+		obj string
+		gid *types.RaftGroupId
+	)
+	if pxy, err = withoutDefaultGroupName(n); err != nil {
+		return nil, err
+	}
+	if obj, err = objectNameForProxy(n); err != nil {
+		return nil, err
+	}
+	if gid, err = m.createGroupID(ctx, n); err != nil {
+		return nil, err
+	}
+	prxy, err := newProxy(m.bundle, gid, sn, pxy, obj)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
The object name and proxy name creation logic is inside the `proxyFactory` in all clients. Otherwise, we need to move the methods to `proxy`, so there is no need for `proxyFactory` which means the `Subsystem` will directly create proxies. In this case, `Subsystem` will be more complex.